### PR TITLE
MV-423: [MV-MIOpen] Benchmark & Port KLDivLoss

### PR DESCRIPTION
**Pull request by hieule88**
* Added no reduction KLDivLoss forward and backward operation and kernel.
* Added reduction KLDivLoss forward and backward operation and kernel.
* Added driver test and gtest for KLDivLoss.
* New API is guarded by MIOPEN_BETA_API macro.
* Compared to ROCm pytorch: [Link](https://morehio-my.sharepoint.com/:x:/r/personal/kyeonghwan_ryu_moreh_io/_layouts/15/Doc.aspx?sourcedoc=%7B890B0782-D902-4EF7-8A05-35C25CA280D5%7D&file=KLDivLoss_benchmark.xlsx&action=default&mobileredirect=true)

* Average over all cases:

No Reduction: 
|type    |Forward| Backward |
|--------|-------|-------|
|float16 | lower  | 2.33 |
|float32 | lower | 2.05 | 
|bfloat16| lower | 2.22 |

Reduction:
|type    |Forward| Backward |
|--------|-------|-------------|
|float16 | 1.05  | 2.42 |
|float32 | 1.08  | 2.15  |
|bfloat16| 1.02  | 2.32 |

### CODE REVIEW

File: .githooks/pre-commit [hunk 1]
```diff
@@ -39,5 +39,4 @@ do
         echo "$format $file"
         "$format" -i -style=file "$file"
     fi
-done
-
+done
```
Comment (kyeonghwanryu) [inline .githooks/pre-commit]: Please remove irrelevant modifications.

Comment (hieule88) [reply .githooks/pre-commit]: removed it

**Comment by kyeonghwanryu** [comment]
There is no info if in/out tensors are contiguous or not.
It seems you ported only non-cont kernels. Is that right?

**Comment by kyeonghwanryu** [comment]
And it seems all forward kernels are not winning much to current rocm.
Improvement must be at least 1.1. I know it might be frustrated, please remove forward ones.

File: src/kldivloss/problem_description.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+NetworkConfig UnreducedProblemDescription::MakeNetworkConfig() const
+{
+    size_t numel       = GetNtotal();
+    size_t num_batches = inputDesc.GetLengths()[0];
+    size_t num_dims    = inputDesc.GetSize();
+    bool is_log_target = GetLogTarget();
+    auto input_dtype   = inputDesc.GetType();
+
+    std::ostringstream ss;
```
Comment (kyeonghwanryu) [inline src/kldivloss/problem_description.cpp:61]: It should include the strides info.

**Comment by hieule88** [comment]
> There is no info if in/out tensors are contiguous or not. It seems you ported only non-cont kernels. Is that right?

Yes, I found that there are only non-contiguous kernels in modnn, and they can also be applied to contiguous input.

**Comment by hieule88** [comment]
> And it seems all forward kernels are not winning much to current rocm. Improvement must be at least 1.1. I know it might be frustrated, please remove forward ones.

Can I just return the value of IsApplicable as false so I can keep the code for future improvements?

File: src/kldivloss/problem_description.cpp (hunk 2 above)
Comment (hieule88) [reply src/kldivloss/problem_description.cpp:61]: included in [commit ](https://github.com/moreh-dev/MIOpen/pull/19/commits/107e18252d29c9a97aa83bf919a1602c5f26a77b)

**Comment by kyeonghwanryu** [comment]
> > And it seems all forward kernels are not winning much to current rocm. Improvement must be at least 1.1. I know it might be frustrated, please remove forward ones.
> 
> Can I just return the value of IsApplicable as false so I can keep the code for future improvements?

Well I think you can do that, but you'll need to persuade AMD reviewer.
But at least remove the API from miopen.h. It would confuse users.

**Comment by hieule88** [comment]
> > > And it seems all forward kernels are not winning much to current rocm. Improvement must be at least 1.1. I know it might be frustrated, please remove forward ones.
> > 
> > 
> > Can I just return the value of IsApplicable as false so I can keep the code for future improvements?
> 
> Well I think you can do that, but you'll need to persuade AMD reviewer. But at least remove the API from miopen.h. It would confuse users.

I created another branch to store the full implementation and removed the forward part of this PR in [this commit](https://github.com/moreh-dev/MIOpen/pull/19/commits/49041c0b0574dbf8cb8a2eaffbc1a39e00775da1).

File: src/kldivloss/problem_description.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    size_t num_dims    = inputDesc.GetSize();
+    bool is_log_target = GetLogTarget();
+    auto input_dtype   = inputDesc.GetType();
+    auto Si            = inputDesc.GetStrides();
+    auto St            = targetDesc.GetStrides();
+    auto So            = outputDesc.GetStrides();
+
+    std::ostringstream ss;
+
+    ss << "kldivloss_unreduce";
```
Comment (kyeonghwanryu) [inline src/kldivloss/problem_description.cpp:63]: guess it's a typo? should be `unreduced`. Same for line 90.

**Comment by kyeonghwanryu** [comment]
It looks good. All that left are simple typos. I'll close this but please fix them.
Good work.

File: src/kldivloss/problem_description.cpp (hunk 3 above)
Comment (hieule88) [reply src/kldivloss/problem_description.cpp:63]: fixed it

**Pull request closed without merging**